### ISSUE TITLE
Should `config_ordinal` leak into `Config`?

### ISSUE THREAD

**Issue by radcortez** [eclipse/microprofile-config#676]
**Description**

The `config_ordinal` is just a local property to a `ConfigSource` to help with the sort. Does it make sense for `config_ordinal` to be available in `Config.getPropertyNames`? Or to return a value in `Config.getValue` (and derivatives)?

In my opinion I don't think so. As mentioned, `config_ordinal` is associated with a single `ConfigSource` with no relationship with the others. Making a call to `Config.getValue` does not have any meaning. It is just the highest available ordinal  and we don't even know the `ConfigSource`. Same for other methods, like `Config.getPropertyNames` or `Config.getValues`.

**Comment by Emily-Jiang** [comment]
`config_ordinal` is a special property and it is used to indicate the ordinal value of the associated config source as you also noted. It should not be surfaced outside the config source. Besides its value can be retrieved via configsource.getOrdinal().

**Comment by radcortez** [comment]
Ok, then we probably need to add some TCK tests for this.

**Comment by ljnelson** [comment]
I'd like to understand what is being proposed.

Currently `Config.getPropertyNames()` [can return whatever it wants (and still be specification compliant)](https://javadoc.io/static/org.eclipse.microprofile.config/microprofile-config-api/2.0/org/eclipse/microprofile/config/Config.html#getPropertyNames--), yes?

Is there a proposal in this issue to make `config_ordinal` a "magic" property name, such that `someConfigSource.getValue("config_ordinal")` should always return `null`, for all `ConfigSource` implementations?

**Comment by radcortez** [comment]
> Currently `Config.getPropertyNames()` [can return whatever it wants (and still be specification compliant)](https://javadoc.io/static/org.eclipse.microprofile.config/microprofile-config-api/2.0/org/eclipse/microprofile/config/Config.html#getPropertyNames--), yes?

Yes. I guess we could leave it as is, but I was mostly wondering if there is any worth is returning it?

> 
> Is there a proposal in this issue to make `config_ordinal` a "magic" property name, such that `someConfigSource.getValue("config_ordinal")` should always return `null`, for all `ConfigSource` implementations?

I think that `ConfigSource.getValue("config_ordinal")` should return a value, but `Config.getValue("config_ordinal")` should not return one. In practice, `config_ordinal` is a magic property.

**Comment by ljnelson** [comment]
In practice, then, doesn't this mean that a `Config#getPropertyNames()` implementation that wants to do the right thing needs to filter the individual return values of `ConfigSource#getPropertyNames()` to exclude "`config_ordinal`" from the sequence that the caller will iterate over?

**Comment by radcortez** [comment]
Yes.

**Comment by ljnelson** [comment]
So it _is_ a property, because a `ConfigSource` can return a value for it (I went looking to see if I could find text that indicates that a `Config` must expose all and only the property values that its `ConfigSource`s expose, but I couldn't find it, but I believe that is the intention of the overall MicroProfile Config effort, i.e. a `Config` does not add state above and beyond that surfaced by its `ConfigSource`s).  But it is _not_ a property in that a `Config#getValue()` call should not surface it (so with this proposal a `Config` now quietly changes its nature and becomes a state-subtracting entity.  This is all extraordinarily weird IMHO.

**Comment by ljnelson** [comment]
Perhaps the larger issue goes something like this:
* There are some things about configuration that concern the configuration itself.  Let's call this metaconfiguration.
  * (Many people have many opinions on _how_ to express metaconfiguration.  **I don't want to go into those opinions here.**)
* One bit of metaconfiguration that exists somewhat hazily in this specification is the `config_ordinal` property.  It is a magic property in some ill-defined sense that is used for sorting purposes.  Another might be the fact that a `ConfigSource`'s name, apparently, maybe, is used as a tiebreaker in sorting, but no guidance is given nor restrictions placed on what that name must or should be.
* There may come to be other such pseudo-magic metaconfiguration down the road.
* **How should the specification express such metaconfiguration in a general, rigorous manner?**  Currently it has taken a very _ad hoc_ approach.  For example, we have this issue, and we also have the long-drawn-out discussion about how to indicate the removal of a property value versus its emptiness (which confuses metaconfiguration with configuration).  We also have the issue of default configuration values, which one could argue treads on metaconfiguration territory.

**Comment by ljnelson** [comment]
So: back to the concrete details of this very particular issue.  My take is this: either `config_ordinal` is a metaconfiguration artifact, and thus not at all part of the property names and property values represented by the overall configuration system, or it is not.  Clearly it is.  Therefore it should be "invisible" wherever possible.  More rigorously:
* The return value of `configSource.getValue("config_ordinal")` MUST be `null` for any `configSource` at any time.
* The return value of `config.getValue("config_ordinal", anyType)` for any `config` and any `anyType` at any time MUST be a non-`null`, [empty](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#isPresent--) `Optional` (its `isPresent()` method must return `false`).
* The Java `String` equal to `config_ordinal` MUST not be yielded by `Iterator#next()` when iterating over the `Iterable` returned by `Config#getPropertyNames()` at any time.
* The Java `String` equal to `config_ordinal` MUST not be contained by the `Set` of `String`s returned by `configSource.getPropertyNames()`, for any `configSource`, at any time.

**Comment by Emily-Jiang** [comment]
After thinking the issue all over and reading all of the comments, here is my 2nd thought.
The property `config_ordinal` is associated with the embed configsource to indicate its ordinal. It is a reserved property, as stated in my previous comment.

We could stop this property from surfacing out. However, do we gain anything by doing so? I think there is a valid use case to surface this property:

As an end users, I would like to find out the highest ordinal number among my config sources. If someone calls config.getValue("config_ordinal", Integer.class), the highest config_ordinal will be returned. 

Considering the cons/pros with supressing this property (treating this property special) vs. surfacing this property (normal property), I think surfacing this property has less impact to the existing customers. If we suppressing this property, the current config.getValue("config_ordinal", Integer.class) will throw exception.

Since the spec said this `config_ordinal` is a property, I think we are good. Do we want to add a tck to enforce this or leave it as it is? 
Thoughts?

**Comment by ljnelson** [comment]
> The property config_ordinal is associated with the embed configsource to indicate its ordinal.

It is configuration about the configuration system, in other words, and thus metaconfiguration.

> It is a reserved property, as stated in my previous comment.

It is "reserved" because it is metaconfiguration but is mixed in with regular configuration.  That's why it has these seemingly magical characteristics, not because it is "reserved".  (What would be the point of reserving configuration? There could be none.)

I am fine if we decide basically: "Oops, we messed up; `config_ordinal` never really should have been in there, but it is, so we treat it no differently from any other property".

I am _not_ fine if we decide: "Yes, it makes perfect sense for metaconfiguration and configuration to mix like this without any way of telling which is which".

I started this discussion not really agreeing with @radcortez's position and I am coming to the end basically supporting a stronger version of it.

**Comment by radcortez** [comment]
@ljnelson thanks for the detailed proposal. My main point was also to discuss this matter widely and figure out where should we move.

I'm +1 on it.

**Comment by ljnelson** [comment]
I'm (obviously) +1 on it too, but I still want to think about it because it is another breaking change.  It is a _good_ breaking change, but breaking nonetheless.

Thinking bigger: what would the process be within MicroProfile (I should know this as a committer but I don't) to open discussion about larger issues like {waves hands} metaconfiguration in general?  Opening a Github issue seems like a terrible tool for such a thing, and hangouts or meetups seem too transient.  I've seen some excellent "larger" Github issues in the area of config simply be closed and effectively ignored and so I don't want that to happen.

**Comment by radcortez** [comment]
I don't believe there is a strict process for it, but probably the mailing list would raise more awareness and bring more people into the discussion.
